hudi-spark/src/main/java/org/apache/hudi/keygen/CustomKeyGenerator.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.keygen;
+
+import org.apache.hudi.DataSourceWriteOptions;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.config.TypedProperties;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hudi.exception.HoodieDeltaStreamerException;
+import org.apache.hudi.exception.HoodieKeyException;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.noPartition;
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.simple;
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.timestampBased;
+
+/**
+ * This is a generic implementation of KeyGenerator where users can configure record key as a single field or a combination of fields.
+ * Similarly partition path can be configured to have multiple fields or only one field. This class expects value for prop
+ * "hoodie.datasource.write.partitionpath.field" in a specific format. For example:
+ *
+ * properties.put("hoodie.datasource.write.partitionpath.field", "field1:PartitionKeyType1,field2:PartitionKeyType2").
+ *
+ * The complete partition path is created as <value for field1 basis PartitionKeyType1>/<value for field2 basis PartitionKeyType2> and so on.
+ *
+ * Few points to consider:
+ * 1. If you want to customise some partition path field on a timestamp basis, you can use field1:timestampBased
+ * 2. If you simply want to have the value of your configured field in the partition path, use field1:simple
+ * 3. If you want your table to be non partitioned, use field1:noPartition.
```

Comment:
or can they leave the partitionPathFields as empty?  that seems more intuitive?

Reply:
Wanted to have a uniform experience in terms of configuring keyGenerator type here. If you insist, I can change it. Please let me know @vinothchandar

Reply:
If you think, it will improve user experience we can do it. I mean not specifying partition field can default to noPartition and it wont be a surprising outcome for the user, right?

Reply:
I am fine with both ways. Let me change it as you say :)